### PR TITLE
[flutter_tools] refactor devtools handler into its own class

### PR DESCRIPTION
**Pull request by jonahwilliams**
In order to support more throughout testing of the interaction between resident runner and devtools, I've refactored this logic into its own class. Internally, this class never calls any method `unwaited`, so it is much easier to test reliably. This also reduces the dependencies between this and the resident runner to the public API, removing the somewhat awkward state/null checking needed to verify that the devtools launcher was non null.

Also fixes an issue where the vm_service could be null.

### CODE REVIEW

**Comment by jonahwilliams** [comment]
Adding @kenzieschmoll and @Hixie since I think you might still be working on some updates to this behavior.

File: packages/flutter_tools/lib/src/resident_devtools_handler.dart [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+class ResidentDevtoolsHandler {
+  ResidentDevtoolsHandler(this._devToolsLauncher, this._residentRunner, this._logger);
+
+  final DevtoolsLauncher _devToolsLauncher;
+  final ResidentRunner _residentRunner;
+  final Logger _logger;
+  bool _shutdown = false;
+  bool _served = false;
+
+  DevToolsServerAddress activeDevToolsServer() =>  _devToolsLauncher?.activeDevToolsServer;
```
Comment (kenzieschmoll) [inline packages/flutter_tools/lib/src/resident_devtools_handler.dart:28]: why not a getter instead?

File: packages/flutter_tools/test/general.shard/resident_devtools_handler_test.dart [hunk 2]
```diff
@@ -0,0 +1,232 @@
...
+  });
+
+  testWithoutContext('can serveAndAnnounceDevTools with attached device does not fail on null vm service', () async {
+    final ResidentDevtoolsHandler handler = ResidentDevtoolsHandler(
+      FakeDevtoolsLauncher()..activeDevToolsServer = DevToolsServerAddress('localhost', 8080),
+      FakeResidentRunner(),
+      BufferLogger.test(),
+    );
+
+    // VM Service is intentionall null
```
Comment (kenzieschmoll) [inline packages/flutter_tools/test/general.shard/resident_devtools_handler_test.dart:118]: nit: typo "intentionally"

**Review by kenzieschmoll — APPROVED** [review]

File: packages/flutter_tools/lib/src/resident_devtools_handler.dart [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+import 'dart:async';
+
+import 'package:meta/meta.dart';
+import 'package:vm_service/vm_service.dart' as vm_service;
+
+import 'base/logger.dart';
+import 'resident_runner.dart';
+import 'vmservice.dart';
+
+
```
[the remark this answers is not in the document]
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/resident_devtools_handler.dart]: Fixed

File: packages/flutter_tools/lib/src/resident_devtools_handler.dart (hunk 1 above)
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/resident_devtools_handler.dart]: Good call, fixed

File: packages/flutter_tools/test/general.shard/resident_devtools_handler_test.dart (hunk 2 above)
Comment (jonahwilliams) [reply packages/flutter_tools/test/general.shard/resident_devtools_handler_test.dart]: D'oh

**Pull request merged**